GGML inspired int8 MM Metal shader

larryliu0820 opened the pull request:
## Context

This PR ported GGML int8 per channel matrix multiplication and matrix vector multiplication metal shaders into ATen library.
llama.cpp LICENSE: https://github.com/ggerganov/llama.cpp/blob/master/LICENSE

## Key Changes

Made the following changes to the original code:

* Memory layout of weight and scales is different than llama.cpp.
* Weight dequantization (scales multiplication) is done after MM is finished.
* Following PyTorch naming convention (M, K, N and assuming row major).

## Benchmark

When M = 1, mv shader improves existing ATen int8mm by 40%.
When M > 4, mm shader outperforms existing ATen int8mm up to 10x for a large M, as show blow.
![image](https://github.com/pytorch/pytorch/assets/8188269/fd9eff71-c538-4263-a7b5-f96fe479ae9d)

Hence the kernel chooses different shaders based on M.

## Test Plan

Tests are passing:
```
❯ python test/test_mps.py -v -k _int8_
/Users/larryliu/CLionProjects/pytorch/venv/lib/python3.8/site-packages/torchvision/io/image.py:13: UserWarning: Failed to load image Python extension: 'dlopen(/Users/larryliu/CLionProjects/pytorch/venv/lib/python3.8/site-packages/torchvision/image.so, 0x0006): Symbol not found: __ZN3c1017RegisterOperatorsD1Ev
  Referenced from: <A770339A-37C9-36B2-84FE-4125FBE26FD6> /Users/larryliu/CLionProjects/pytorch/venv/lib/python3.8/site-packages/torchvision/image.so
  Expected in:     <5749F98A-0A0C-3F89-9CBF-277B3C8EA00A> /Users/larryliu/CLionProjects/pytorch/torch/lib/libtorch_cpu.dylib'If you don't plan on using image functionality from `torchvision.io`, you can ignore this warning. Otherwise, there might be something wrong with your environment. Did you have `libjpeg` or `libpng` installed before building `torchvision` from source?
  warn(
test__int8_mm_m_1_k_32_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_1_k_32_n_64_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_1_k_64_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_1_k_64_n_64_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_32_k_32_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_32_k_32_n_64_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_32_k_64_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_32_k_64_n_64_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_64_k_32_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_64_k_32_n_64_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_64_k_64_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_64_k_64_n_64_mps (__main__.TestLinalgMPSMPS) ... ok

----------------------------------------------------------------------
Ran 12 tests in 1.180s

OK
```

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #127646

larryliu0820 committed:
GGML inspired int8 MM Metal shader

[ghstack-poisoned]

larryliu0820 committed:
Update on "GGML inspired int8 MM Metal shader"

TODO: Add a bunch of comments for the algorithm.

Tests are passing:
```
❯ python test/test_mps.py -v -k _int8_
/Users/larryliu/CLionProjects/pytorch/venv/lib/python3.8/site-packages/torchvision/io/image.py:13: UserWarning: Failed to load image Python extension: 'dlopen(/Users/larryliu/CLionProjects/pytorch/venv/lib/python3.8/site-packages/torchvision/image.so, 0x0006): Symbol not found: __ZN3c1017RegisterOperatorsD1Ev
  Referenced from: <A770339A-37C9-36B2-84FE-4125FBE26FD6> /Users/larryliu/CLionProjects/pytorch/venv/lib/python3.8/site-packages/torchvision/image.so
  Expected in:     <5749F98A-0A0C-3F89-9CBF-277B3C8EA00A> /Users/larryliu/CLionProjects/pytorch/torch/lib/libtorch_cpu.dylib'If you don't plan on using image functionality from `torchvision.io`, you can ignore this warning. Otherwise, there might be something wrong with your environment. Did you have `libjpeg` or `libpng` installed before building `torchvision` from source?
  warn(
test__int8_mm_m_1_k_32_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_1_k_32_n_64_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_1_k_64_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_1_k_64_n_64_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_32_k_32_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_32_k_32_n_64_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_32_k_64_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_32_k_64_n_64_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_64_k_32_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_64_k_32_n_64_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_64_k_64_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_64_k_64_n_64_mps (__main__.TestLinalgMPSMPS) ... ok

----------------------------------------------------------------------
Ran 12 tests in 1.180s

OK
```




[ghstack-poisoned]

larryliu0820 committed:
Update on "GGML inspired int8 MM Metal shader"

TODO: Add a bunch of comments for the algorithm.

Tests are passing:
```
❯ python test/test_mps.py -v -k _int8_
/Users/larryliu/CLionProjects/pytorch/venv/lib/python3.8/site-packages/torchvision/io/image.py:13: UserWarning: Failed to load image Python extension: 'dlopen(/Users/larryliu/CLionProjects/pytorch/venv/lib/python3.8/site-packages/torchvision/image.so, 0x0006): Symbol not found: __ZN3c1017RegisterOperatorsD1Ev
  Referenced from: <A770339A-37C9-36B2-84FE-4125FBE26FD6> /Users/larryliu/CLionProjects/pytorch/venv/lib/python3.8/site-packages/torchvision/image.so
  Expected in:     <5749F98A-0A0C-3F89-9CBF-277B3C8EA00A> /Users/larryliu/CLionProjects/pytorch/torch/lib/libtorch_cpu.dylib'If you don't plan on using image functionality from `torchvision.io`, you can ignore this warning. Otherwise, there might be something wrong with your environment. Did you have `libjpeg` or `libpng` installed before building `torchvision` from source?
  warn(
test__int8_mm_m_1_k_32_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_1_k_32_n_64_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_1_k_64_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_1_k_64_n_64_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_32_k_32_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_32_k_32_n_64_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_32_k_64_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_32_k_64_n_64_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_64_k_32_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_64_k_32_n_64_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_64_k_64_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_64_k_64_n_64_mps (__main__.TestLinalgMPSMPS) ... ok

----------------------------------------------------------------------
Ran 12 tests in 1.180s

OK
```




[ghstack-poisoned]

larryliu0820 committed:
Update on "GGML inspired int8 MM Metal shader"

TODO: Add a bunch of comments for the algorithm.

Tests are passing:
```
❯ python test/test_mps.py -v -k _int8_
/Users/larryliu/CLionProjects/pytorch/venv/lib/python3.8/site-packages/torchvision/io/image.py:13: UserWarning: Failed to load image Python extension: 'dlopen(/Users/larryliu/CLionProjects/pytorch/venv/lib/python3.8/site-packages/torchvision/image.so, 0x0006): Symbol not found: __ZN3c1017RegisterOperatorsD1Ev
  Referenced from: <A770339A-37C9-36B2-84FE-4125FBE26FD6> /Users/larryliu/CLionProjects/pytorch/venv/lib/python3.8/site-packages/torchvision/image.so
  Expected in:     <5749F98A-0A0C-3F89-9CBF-277B3C8EA00A> /Users/larryliu/CLionProjects/pytorch/torch/lib/libtorch_cpu.dylib'If you don't plan on using image functionality from `torchvision.io`, you can ignore this warning. Otherwise, there might be something wrong with your environment. Did you have `libjpeg` or `libpng` installed before building `torchvision` from source?
  warn(
test__int8_mm_m_1_k_32_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_1_k_32_n_64_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_1_k_64_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_1_k_64_n_64_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_32_k_32_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_32_k_32_n_64_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_32_k_64_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_32_k_64_n_64_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_64_k_32_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_64_k_32_n_64_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_64_k_64_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_64_k_64_n_64_mps (__main__.TestLinalgMPSMPS) ... ok

----------------------------------------------------------------------
Ran 12 tests in 1.180s

OK
```




[ghstack-poisoned]

larryliu0820 committed:
Update on "GGML inspired int8 MM Metal shader"

TODO: Add a bunch of comments for the algorithm.

Tests are passing:
```
❯ python test/test_mps.py -v -k _int8_
/Users/larryliu/CLionProjects/pytorch/venv/lib/python3.8/site-packages/torchvision/io/image.py:13: UserWarning: Failed to load image Python extension: 'dlopen(/Users/larryliu/CLionProjects/pytorch/venv/lib/python3.8/site-packages/torchvision/image.so, 0x0006): Symbol not found: __ZN3c1017RegisterOperatorsD1Ev
  Referenced from: <A770339A-37C9-36B2-84FE-4125FBE26FD6> /Users/larryliu/CLionProjects/pytorch/venv/lib/python3.8/site-packages/torchvision/image.so
  Expected in:     <5749F98A-0A0C-3F89-9CBF-277B3C8EA00A> /Users/larryliu/CLionProjects/pytorch/torch/lib/libtorch_cpu.dylib'If you don't plan on using image functionality from `torchvision.io`, you can ignore this warning. Otherwise, there might be something wrong with your environment. Did you have `libjpeg` or `libpng` installed before building `torchvision` from source?
  warn(
test__int8_mm_m_1_k_32_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_1_k_32_n_64_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_1_k_64_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_1_k_64_n_64_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_32_k_32_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_32_k_32_n_64_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_32_k_64_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_32_k_64_n_64_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_64_k_32_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_64_k_32_n_64_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_64_k_64_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_64_k_64_n_64_mps (__main__.TestLinalgMPSMPS) ... ok

----------------------------------------------------------------------
Ran 12 tests in 1.180s

OK
```




[ghstack-poisoned]

larryliu0820 committed:
Update on "GGML inspired int8 MM Metal shader"

## Context

This PR ported GGML matrix multiplication and matrix vector multiplication metal shaders into ATen library.

llama.cpp LICENSE: https://github.com/ggerganov/llama.cpp/blob/master/LICENSE

## Key Changes

Made the following changes to the original code:

* Memory layout of weight and scales is different than llama.cpp.
* Weight dequantization (scales multiplication) is done after MM is finished.
* Following PyTorch naming convention (M, K, N and assuming row major).

## Benchmark

When M = 1, mv shader improves existing ATen int8mm by 40%.
When M > 4, mm shader outperforms existing ATen int8mm.
![image](https://github.com/pytorch/pytorch/assets/8188269/fd9eff71-c538-4263-a7b5-f96fe479ae9d)

Hence the kernel chooses different shaders based on M.

## Test Plan

Tests are passing:
```
❯ python test/test_mps.py -v -k _int8_
/Users/larryliu/CLionProjects/pytorch/venv/lib/python3.8/site-packages/torchvision/io/image.py:13: UserWarning: Failed to load image Python extension: 'dlopen(/Users/larryliu/CLionProjects/pytorch/venv/lib/python3.8/site-packages/torchvision/image.so, 0x0006): Symbol not found: __ZN3c1017RegisterOperatorsD1Ev
  Referenced from: <A770339A-37C9-36B2-84FE-4125FBE26FD6> /Users/larryliu/CLionProjects/pytorch/venv/lib/python3.8/site-packages/torchvision/image.so
  Expected in:     <5749F98A-0A0C-3F89-9CBF-277B3C8EA00A> /Users/larryliu/CLionProjects/pytorch/torch/lib/libtorch_cpu.dylib'If you don't plan on using image functionality from `torchvision.io`, you can ignore this warning. Otherwise, there might be something wrong with your environment. Did you have `libjpeg` or `libpng` installed before building `torchvision` from source?
  warn(
test__int8_mm_m_1_k_32_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_1_k_32_n_64_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_1_k_64_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_1_k_64_n_64_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_32_k_32_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_32_k_32_n_64_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_32_k_64_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_32_k_64_n_64_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_64_k_32_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_64_k_32_n_64_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_64_k_64_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_64_k_64_n_64_mps (__main__.TestLinalgMPSMPS) ... ok

----------------------------------------------------------------------
Ran 12 tests in 1.180s

OK
```




[ghstack-poisoned]

iseeyuan commented:
In the summary, "This PR ported GGML **int8 per channel** matrix multiplication and matrix vector multiplication metal shaders into ATen library."

iseeyuan commented:
"When M > 4, mm shader outperforms existing ATen int8mm **up to 10x for a large M**, as show blow", or something like that.

iseeyuan commented on aten/src/ATen/native/mps/operations/Quantized.mm:
```diff
@@ -295,15 +749,32 @@ Tensor _weight_int8pack_mm_mps(const Tensor& A, const Tensor& B, const Tensor& s
+      if (M == 1) {
+        kernel = fmt::format("int8pack_mv_{}", scalarToMetalTypeString(A));
+      } else if (M < 4) {
+        kernel = fmt::format("int8pack_mm_{}", scalarToMetalTypeString(A));
```
So for M<4 we keep using ATen kernel?

larryliu0820 replied:
Yes

malfet replied:
How much perf would we loose if we are to use mv kernel for those cases? (Just call it M times)
It would be good to have two kernels (mv + mm) rather than 3

larryliu0820 replied:
I can do a follow-up to clean this up.

malfet reviewed (commented):
Can you please add test for all 3 kernels? (Right now there is a test for M=1, as well as for M>4, but there are no tests for say M=2 (the third kernel)

Also, how big the the perf difference between mv op vs small_mm one? (I.e. I would much prefer to have 2 kernels rather than 3)

malfet reviewed (approved):
I see a good perf increase(from 16.4 to 17.8 tokens/sec on M2Pro), so let's land this

malfet commented on aten/src/ATen/native/mps/operations/Quantized.mm:
```diff
@@ -204,6 +204,460 @@ kernel void int4pack_mm(
+
+    // 8192 for sa, 4096 for sb
+    threadgroup T      * sa = (threadgroup T      *)(shared_memory);
+    threadgroup bfloat * sb = (threadgroup bfloat *)(shared_memory + 8192);
```
Hmm, though this would not compile on Ventura. But also, why it must be of bfloat type always?

larryliu0820 replied:
I'll fix this

swolchok commented on aten/src/ATen/native/mps/operations/Quantized.mm:
```diff
@@ -204,6 +204,460 @@ kernel void int4pack_mm(
+ *
+ * Matrix Multiplication Algorithm:
+ * 1. Load A and B blocks (32x32 and 64x32 respectively) into shared memory.
+ * 2. In 4 simdgroups, calculate the outer product of the loaded blocks. Each simdgroup produces a 2x4 8x8 result.
```
wikipedia is unclear about the meaning of "outer product" as it applies to matrices. are you referring to what they're calling Kronecker product?

swolchok commented on aten/src/ATen/native/mps/operations/Quantized.mm:
```diff
@@ -204,6 +204,460 @@ kernel void int4pack_mm(
+    uint32_t M = sizes.x; // M
+    uint32_t K = sizes.y; // K
+    uint32_t N = sizes.z; // N
+    uint32_t nb00 = sizeof(W);
+    uint32_t nb01 = nb00 * K;
+    uint32_t nb10 = sizeof(T);
+    uint32_t nb11 = nb10 * K;
```
this GGML notation is unreadable IMO, especially since this is PyTorch -- is it in PyTorch order (in which case it doesn't match the convention for the codebase that originated the notation) or GGML order (in which case it doesn't match PyTorch)?

swolchok commented on aten/src/ATen/native/mps/operations/Quantized.mm:
```diff
@@ -204,6 +204,460 @@ kernel void int4pack_mm(
+    uint32_t nb10 = sizeof(T);
+    uint32_t nb11 = nb10 * K;
+
+    // 8192 for sa, 4096 for sb
+    threadgroup T      * sa = (threadgroup T      *)(shared_memory);
+    threadgroup bfloat * sb = (threadgroup bfloat *)(shared_memory + 8192);
```
what does the "s" stand for? shared? why can't we write that out?

swolchok commented on aten/src/ATen/native/mps/operations/Quantized.mm:
```diff
@@ -204,6 +204,460 @@ kernel void int4pack_mm(
+    threadgroup T      * sa = (threadgroup T      *)(shared_memory);
+    threadgroup bfloat * sb = (threadgroup bfloat *)(shared_memory + 8192);
+
+    const uint r0 = tgpig.y;
+    const uint r1 = tgpig.x;
```
I have no basis for determining what an "r" is in this context

swolchok commented on aten/src/ATen/native/mps/operations/Quantized.mm:
```diff
@@ -204,6 +204,460 @@ kernel void int4pack_mm(
+    const uint r0 = tgpig.y;
+    const uint r1 = tgpig.x;
+
+    // if this block is of 64x32 shape or smaller
```
then what?

swolchok commented on aten/src/ATen/native/mps/operations/Quantized.mm:
```diff
@@ -204,6 +204,460 @@ kernel void int4pack_mm(
+    short n_rows_B = (N - r0 * BLOCK_SIZE_N < BLOCK_SIZE_N) ? (N - r0 * BLOCK_SIZE_N) : BLOCK_SIZE_N;
+
+    // a thread shouldn't load data outside of the matrix
+    short thread_row_A = ((short)tiitg/THREAD_PER_ROW_A) < n_rows_A
+        ? ((short)tiitg/THREAD_PER_ROW_A)
+        : n_rows_A - 1;
```
can this be simplified to `((short)tiith / THREAD_PER_ROW_A) <= n_rows_A - 1 ?: n_rows_A - 1`, or perhaps an expression with min()?

malfet replied:
[Elvis operator](https://en.wikipedia.org/wiki/Elvis_operator) is not part of C/C++ standard, nor do I see it mentioned in Metal spec (though it works)

swolchok commented on aten/src/ATen/native/mps/operations/Quantized.mm:
```diff
@@ -204,6 +204,460 @@ kernel void int4pack_mm(
+        ? ((short)tiitg/THREAD_PER_ROW_B)
+        : n_rows_B - 1;
+
+    Tsimd8x8 ma[2]; // input
+    simdgroup_bfloat8x8 mb[4]; // weight
```
what is an "m"? I could guess "matrix", but why do I have to?

swolchok commented on aten/src/ATen/native/mps/operations/Quantized.mm:
```diff
@@ -204,6 +204,460 @@ kernel void int4pack_mm(
+Load weight and input into shared memory:
+
+             K
+  ┌────────────────────────┐              8192 (A)           4096 (B)
```
I am mystified as to the origin and relevance of the 8192 and 4096 constants

swolchok commented on aten/src/ATen/native/mps/operations/Quantized.mm:
```diff
@@ -204,6 +204,460 @@ kernel void int4pack_mm(
+        for (short i = 0; i < 16; i++) {
+            bfloat weight = *(b_ptr + i);
+            // for example, tiitg 32, i 12 -> 0 + 1 = 1, it needs to work on sg mat grid row 1
+            short sg_mat_grid_row_index = (tiitg % THREAD_PER_ROW_B) * THREAD_PER_ROW_B + i / 8;
```
does "sg" stand for "simdgroup"? can we just write "simdgroup"?

swolchok replied:
(and why is this a different "s" from the "s" in "sb" and "sa")

swolchok commented on aten/src/ATen/native/mps/operations/Quantized.mm:
```diff
@@ -204,6 +204,460 @@ kernel void int4pack_mm(
+    for (int i = 0; i < 8; i++) {
+        int block_start = 4 * 8 * (sgitg & 1) + (i % 4) * 8;
+        threadgroup float * temp_scale = (threadgroup float *)sb + block_start;
+        threadgroup float * scale_itr = temp_scale;
```
what is an "itr"? "iterator"? ("iter" would be OK per longstanding programming convention)

swolchok commented on aten/src/ATen/native/mps/operations/Quantized.mm:
```diff
@@ -204,6 +204,460 @@ kernel void int4pack_mm(
+  │                        │   │++++++++++++++++++++++++│++++++++++++│
+  │                        │   └────────────────────────┴────────────┘
+  │                        │
+  │32                      │
+  ├──┬──┬──────────────────┤                    K
+  │++│  │                  │        ┌────────────────────────┐
+64│++│  │  ...             │        │                        │
```
can `BLOCK_SIZE_M = 32`& `BLOCK_SIZE_N = 64` appear in the diagram so that it is easier to track the code and the diagram together?

swolchok commented on aten/src/ATen/native/mps/operations/Quantized.mm:
```diff
@@ -204,6 +204,460 @@ kernel void int4pack_mm(
+  │++│  │                  │        │                        │
+  ├──┴──┴──────────────────┤        │                        │
+  │                        │        │                        │
+  │      ───────────►      │        │32                      │
+  │       for loop         │        ├──┬──┬──────────────────┤
+  │                        │      32│++│  │ ...              │
```
ditto BLOCK_SIZE constants

swolchok commented on aten/src/ATen/native/mps/operations/Quantized.mm:
```diff
@@ -204,6 +204,460 @@ kernel void int4pack_mm(
+            // same example, sg mat grid col index: 32 / 2 / 8 = 2, so currently need to work with sg mat at (1, 2)
+            short sg_mat_grid_col_index = tiitg / THREAD_PER_ROW_B / 8;
+            // now inside sg mat, which index to write to? starting point is SG_MAT_SIZE * sg_mat_offset
+            short row_offset = i & 7;
```
nit: `i % 8` really should get you the same result from any reasonable compiler

malfet replied:
To be frank, I prefer `n & mask` to `n % (mask+1)`, because I remember that failed to do this optimization

swolchok commented on aten/src/ATen/native/mps/operations/Quantized.mm:
```diff
@@ -204,6 +204,460 @@ kernel void int4pack_mm(
+        threadgroup_barrier(mem_flags::mem_threadgroup);
+
+        // load matrices from threadgroup memory and conduct outer products
+        threadgroup T      * lsma = (sa + THREAD_MAT_M * SG_MAT_SIZE * (sgitg / 2));
+        threadgroup bfloat * lsmb = (sb + THREAD_MAT_N * SG_MAT_SIZE * (sgitg % 2));
```
no idea what "lsm" stands for

swolchok commented on aten/src/ATen/native/mps/operations/Quantized.mm:
```diff
@@ -204,6 +204,460 @@ kernel void int4pack_mm(
+#define N_DST 4        // each SIMD group works on 4 rows
+#define N_SIMDGROUP 2  // number of SIMD groups in a thread group
+#define N_SIMDWIDTH 32 // assuming SIMD group size is 32
+#define QK8_0 32       // 32 weights in each packed weight in ggml block_q8_0 type.
```
we aren't using the GGML block_q8_0 type, so why is this here?

swolchok reviewed (commented):
notation is very terse. Is it really necessary not to spell things out?

Bunch of inline comments highlighting specific places I can't read at a glance

larryliu0820 committed:
Update on "GGML inspired int8 MM Metal shader"

## Context

This PR ported GGML int8 per channel matrix multiplication and matrix vector multiplication metal shaders into ATen library.
llama.cpp LICENSE: https://github.com/ggerganov/llama.cpp/blob/master/LICENSE

## Key Changes

Made the following changes to the original code:

* Memory layout of weight and scales is different than llama.cpp.
* Weight dequantization (scales multiplication) is done after MM is finished.
* Following PyTorch naming convention (M, K, N and assuming row major).

## Benchmark

When M = 1, mv shader improves existing ATen int8mm by 40%.
When M > 4, mm shader outperforms existing ATen int8mm up to 10x for a large M, as show blow.
![image](https://github.com/pytorch/pytorch/assets/8188269/fd9eff71-c538-4263-a7b5-f96fe479ae9d)

Hence the kernel chooses different shaders based on M.

## Test Plan

Tests are passing:
```
❯ python test/test_mps.py -v -k _int8_
/Users/larryliu/CLionProjects/pytorch/venv/lib/python3.8/site-packages/torchvision/io/image.py:13: UserWarning: Failed to load image Python extension: 'dlopen(/Users/larryliu/CLionProjects/pytorch/venv/lib/python3.8/site-packages/torchvision/image.so, 0x0006): Symbol not found: __ZN3c1017RegisterOperatorsD1Ev
  Referenced from: <A770339A-37C9-36B2-84FE-4125FBE26FD6> /Users/larryliu/CLionProjects/pytorch/venv/lib/python3.8/site-packages/torchvision/image.so
  Expected in:     <5749F98A-0A0C-3F89-9CBF-277B3C8EA00A> /Users/larryliu/CLionProjects/pytorch/torch/lib/libtorch_cpu.dylib'If you don't plan on using image functionality from `torchvision.io`, you can ignore this warning. Otherwise, there might be something wrong with your environment. Did you have `libjpeg` or `libpng` installed before building `torchvision` from source?
  warn(
test__int8_mm_m_1_k_32_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_1_k_32_n_64_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_1_k_64_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_1_k_64_n_64_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_32_k_32_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_32_k_32_n_64_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_32_k_64_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_32_k_64_n_64_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_64_k_32_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_64_k_32_n_64_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_64_k_64_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_64_k_64_n_64_mps (__main__.TestLinalgMPSMPS) ... ok

----------------------------------------------------------------------
Ran 12 tests in 1.180s

OK
```




[ghstack-poisoned]

larryliu0820 committed:
Update on "GGML inspired int8 MM Metal shader"

## Context

This PR ported GGML int8 per channel matrix multiplication and matrix vector multiplication metal shaders into ATen library.
llama.cpp LICENSE: https://github.com/ggerganov/llama.cpp/blob/master/LICENSE

## Key Changes

Made the following changes to the original code:

* Memory layout of weight and scales is different than llama.cpp.
* Weight dequantization (scales multiplication) is done after MM is finished.
* Following PyTorch naming convention (M, K, N and assuming row major).

## Benchmark

When M = 1, mv shader improves existing ATen int8mm by 40%.
When M > 4, mm shader outperforms existing ATen int8mm up to 10x for a large M, as show blow.
![image](https://github.com/pytorch/pytorch/assets/8188269/fd9eff71-c538-4263-a7b5-f96fe479ae9d)

Hence the kernel chooses different shaders based on M.

## Test Plan

Tests are passing:
```
❯ python test/test_mps.py -v -k _int8_
/Users/larryliu/CLionProjects/pytorch/venv/lib/python3.8/site-packages/torchvision/io/image.py:13: UserWarning: Failed to load image Python extension: 'dlopen(/Users/larryliu/CLionProjects/pytorch/venv/lib/python3.8/site-packages/torchvision/image.so, 0x0006): Symbol not found: __ZN3c1017RegisterOperatorsD1Ev
  Referenced from: <A770339A-37C9-36B2-84FE-4125FBE26FD6> /Users/larryliu/CLionProjects/pytorch/venv/lib/python3.8/site-packages/torchvision/image.so
  Expected in:     <5749F98A-0A0C-3F89-9CBF-277B3C8EA00A> /Users/larryliu/CLionProjects/pytorch/torch/lib/libtorch_cpu.dylib'If you don't plan on using image functionality from `torchvision.io`, you can ignore this warning. Otherwise, there might be something wrong with your environment. Did you have `libjpeg` or `libpng` installed before building `torchvision` from source?
  warn(
test__int8_mm_m_1_k_32_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_1_k_32_n_64_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_1_k_64_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_1_k_64_n_64_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_32_k_32_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_32_k_32_n_64_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_32_k_64_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_32_k_64_n_64_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_64_k_32_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_64_k_32_n_64_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_64_k_64_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_64_k_64_n_64_mps (__main__.TestLinalgMPSMPS) ... ok

----------------------------------------------------------------------
Ran 12 tests in 1.180s

OK
```




[ghstack-poisoned]

larryliu0820 commented:
@pytorchbot merge

malfet commented:
@pytorchbot merge -r

malfet commented on aten/src/ATen/native/mps/operations/Quantized.mm:
```diff
@@ -204,6 +204,458 @@ kernel void int4pack_mm(
+
+    // shared memory for A and B
+    threadgroup T    * shared_memory_A = (threadgroup T    *)(shared_memory);
+    threadgroup half * shared_memory_B = (threadgroup half *)(shared_memory + 8192);
```
Two remarks: - all offsets should be expressed as number of elements multiplied by sizeof elemen
Let's keep A and B types the same
```suggestion
    threadgroup T * shared_memory_B = (threadgroup half *)(shared_memory + 4096 * sizeof(T));
```

larryliu0820 replied:
Added comment, as using `half` here is good for perf (8% improvement) comparing with `bfloat`.

larryliu0820 committed:
Update on "GGML inspired int8 MM Metal shader"

## Context

This PR ported GGML int8 per channel matrix multiplication and matrix vector multiplication metal shaders into ATen library.
llama.cpp LICENSE: https://github.com/ggerganov/llama.cpp/blob/master/LICENSE

## Key Changes

Made the following changes to the original code:

* Memory layout of weight and scales is different than llama.cpp.
* Weight dequantization (scales multiplication) is done after MM is finished.
* Following PyTorch naming convention (M, K, N and assuming row major).

## Benchmark

When M = 1, mv shader improves existing ATen int8mm by 40%.
When M > 4, mm shader outperforms existing ATen int8mm up to 10x for a large M, as show blow.
![image](https://github.com/pytorch/pytorch/assets/8188269/fd9eff71-c538-4263-a7b5-f96fe479ae9d)

Hence the kernel chooses different shaders based on M.

## Test Plan

Tests are passing:
```
❯ python test/test_mps.py -v -k _int8_
/Users/larryliu/CLionProjects/pytorch/venv/lib/python3.8/site-packages/torchvision/io/image.py:13: UserWarning: Failed to load image Python extension: 'dlopen(/Users/larryliu/CLionProjects/pytorch/venv/lib/python3.8/site-packages/torchvision/image.so, 0x0006): Symbol not found: __ZN3c1017RegisterOperatorsD1Ev
  Referenced from: <A770339A-37C9-36B2-84FE-4125FBE26FD6> /Users/larryliu/CLionProjects/pytorch/venv/lib/python3.8/site-packages/torchvision/image.so
  Expected in:     <5749F98A-0A0C-3F89-9CBF-277B3C8EA00A> /Users/larryliu/CLionProjects/pytorch/torch/lib/libtorch_cpu.dylib'If you don't plan on using image functionality from `torchvision.io`, you can ignore this warning. Otherwise, there might be something wrong with your environment. Did you have `libjpeg` or `libpng` installed before building `torchvision` from source?
  warn(
test__int8_mm_m_1_k_32_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_1_k_32_n_64_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_1_k_64_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_1_k_64_n_64_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_32_k_32_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_32_k_32_n_64_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_32_k_64_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_32_k_64_n_64_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_64_k_32_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_64_k_32_n_64_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_64_k_64_n_32_mps (__main__.TestLinalgMPSMPS) ... ok
test__int8_mm_m_64_k_64_n_64_mps (__main__.TestLinalgMPSMPS) ... ok

----------------------------------------------------------------------
Ran 12 tests in 1.180s

OK
```




[ghstack-poisoned]

larryliu0820 commented:
@pytorchbot merge

larryliu0820 commented:
@pytorchbot merge -i "lintrunner-noclang is failing on trunk"

larryliu0820 commented:
lintrunner-noclang is failing on trunk. Ignoring

larryliu0820 commented:
@pytorchbot merge -i

larryliu0820 commented:
@pytorchbot merge -i
